### ISSUE TITLE
Orthogonalize innermost tool infrastructure from Core

### ISSUE THREAD

**Issue by DavidPoliakoff** [kokkos/kokkos#4327]
We have interest from other projects in using the Tools callbacks. These projects probably wouldn't be willing to eat Kokkos as a whole. I'm proposing that we make

```c++
#include <impl/SomeKokkosToolsHeader.hpp> // doesn't include anything from Core
Kokkos::Tools::initialize(argc, argv);
Kokkos::Tools::pushRegion("dogs");
Kokkos::Tools::popRegion();
Kokkos::Tools::finalize();
```

A valid program.

**Comment by mhoemmen** [comment]
@DavidPoliakoff I have an immediate use case for this.

**Comment by DavidPoliakoff** [comment]
I'll attempt to prioritize that, thanks for letting me know. FYI, @sameershende is the original requestor on this

**Comment by DavidPoliakoff** [comment]
I took a stab at implementing this, and really there's only two problems

1. How to determine whether LIBDL is enabled (for now: just assuming yes)
2. How to handle initialization

I had some thoughts on (2). Right now, the way Kokkos initialization goes is something like this

```c++
void Kokkos::initialize(int argc, char* argv[]){
  InitArguments everything_kokkos_knows_how_to_initialize = magic_parser(argc, argv);
  Kokkos::initialize_backends(everything_kokkos_knows_how_to_initialize);
  Kokkos::initialize_profiling(everything_kokkos_knows_how_to_initialize);
  // more lines of this ilk
}
```

The problem, of course, is that InitArguments is a part of Core, and so we can't talk about an InitArguments inside of Tools. Problem Two: users call Kokkos::initialize with InitArguments. What I'd like to swap this to, would be something like this:

```c++

ns Kokkos {
ns Tools {
  struct InitArguments {
    std::string profileLibrary;
    // etc
  };
} // Tools
ns Backends {
  struct InitArguments {
    int num_threads;
    // etc
  };
} // Backends
struct NewInitArguments {
  NewInitArguments(InitArguments) {
    // parse out the fields of the old InitArguments struct
  }
  Tools::InitArguments tools;
  Backends::InitArguments backends;
}

}


void Kokkos::initialize(NewInitArguments init){
  Kokkos::Backends::Impl::initialize(init.backends); 
  Kokkos::Tools::initialize(init.tools); // probably *not* impl
}
void Kokkos::initialize(InitArguments old_init) { // this is the original InitArguments guy
   Kokkos::Tools::NewInitArguments init(old_init);
   initialize(init);
}
void Kokkos::initialize(int argc, char* argv[]){
  Kokkos::Tools::InitArguments tool_init_arguments = Kokkos::Tools::parse_arguments(argc, argv);
  Kokkos::Backends::InitArguments backend_init_arguments = Kokkos::Backends::parse_arguments(argc, argv);
  Kokkos::NewInitArguments init {tool_init_arguments, backend_init_arguments};
  initialize(init);
}
```

What this would let Kokkos do, would be to better orthogonalize the various parts of Kokkos, argument parsing _could_ be split up if we wanted. What this would let _Tools_ do, would be to completely allow us to be initialized independently of Kokkos. Once we have that, we have everything Sameer and Mark need, either we split Tools off into its own directory/repo, or they can just copy out the set of files I tell them they need to support Tools.

**Comment by crtrott** [comment]
I would propose you add a new private member to Kokkos::InitArguments, which is the Profiling InitArguments. 
We deprecate the direct tools members 
```
  bool disable_warnings;
  bool tune_internals;
  bool tool_help        = false;
  std::string tool_lib  = {};
  std::string tool_args = {};
```

Until they are gone we check in initialize that you don't have set both them and the nested tools struct. 
Add a constructor which can also add that new guy, add a private flag whcih gets set if that internal thing was set via a consturctor (and maybe via a set function??). 

Also initialize in Kokkos is structurally more like:

```
pre_initialize(...);
for(backend in backends) backend.initialize(...);
pose_initialize(...);
```

Now one interesting thing is whether tools should be in the pre or post initialize. Currently I think they are in the post or?
Reason for Pre: catches backend initialization stuff like allocations of scratch buffers etc.
Reason for Post: tool could actually access accelerator data, maybe even just stuff like which GPU is this process using and what properties does it have.

Currently I think the Post stuff is probably better.

**Comment by DavidPoliakoff** [comment]
There's still the problem of needing to allow sub-components of Kokkos to parse the arguments. Without that we're pretty hosed. Is that part kosher?

**Comment by ajpowelsnl** [comment]
Morning - @DavidPoliakoff and @crtrott  -- Would you like to discuss design of this item in our Wed. meeting?

**Comment by DavidPoliakoff** [comment]
We'll see how it goes beating one another up on Slack first :)

**Comment by ajpowelsnl** [comment]
> We'll see how it goes beating one another up on Slack first :)

HA!  LMK if we need to move it into the ThunderDome (Wed. Meeting)

**Comment by PhilMiller** [comment]
Fix released in 3.6
